### ISSUE TITLE
OnnxImportException: Unexpected error while parsing layer 2452 of type Unsqueeze. Only tensors of rank 4 or less are supported, but got rank 5

### ISSUE THREAD

**Issue by jaggernaut007** [Unity-Technologies/barracuda-release#50]
*Stack Trace*

Json: { "input": [ "2435" ], "output": [ "2452" ], "name": "Unsqueeze_173", "opType": "Unsqueeze", "attribute": [ { "name": "axes", "ints": [ "0" ], "type": "INTS" } ] }
  at Unity.Barracuda.ONNXLayout.AxisPermutationsForMappingONNXLayoutToBarracuda (System.Int32 onnxRank, System.String onnxLayout) [0x00028] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXLayout.cs:44 
  at Unity.Barracuda.ONNXLayout.PermuteToBarracuda (System.Int64[] shape, System.String onnxLayout) [0x00005] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXLayout.cs:158 
  at Unity.Barracuda.ONNXLayout.ConvertSymbolicShapeToBarracuda (System.Int64[] onnxShape, System.String onnxLayout) [0x00001] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXLayout.cs:223 
  at Unity.Barracuda.ONNXTensor.Reshape (System.Int64[] onnxShape) [0x00001] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXTensor.cs:184 
  at Unity.Barracuda.ONNXTensor.Unsqueeze (System.Int32[] axes) [0x0003a] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXTensor.cs:237 
  at Unity.Barracuda.ONNXModelImporter.<.ctor>b__14_4 (Unity.Barracuda.ModelBuilder net, Unity.Barracuda.ONNXNodeWrapper node) [0x0000c] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXModelImporter.cs:177 
  at Unity.Barracuda.ONNXModelImporter.ConvertOnnxModel (Onnx.ModelProto onnxModel) [0x00367] in C:\Users\Shreyas\AIMage\12345\Library\PackageCache\com.unity.barracuda@0.7.1-preview\Barracuda\Editor\ONNXModelImporter.cs:1043 

Unity.Barracuda.ONNXModelImporter.Err (Unity.Barracuda.Model model, System.String layerName, System.String message, System.String extendedMessage, System.String debugMessage) (at Library/PackageCache/com.unity.barracuda@0.7.1-preview/Barracuda/Editor/ONNXModelImporter.cs:1358)
Unity.Barracuda.ONNXModelImporter.ConvertOnnxModel (Onnx.ModelProto onnxModel) (at Library/PackageCache/com.unity.barracuda@0.7.1-preview/Barracuda/Editor/ONNXModelImporter.cs:1052)
Unity.Barracuda.ONNXModelImporter.OnImportAsset (UnityEditor.Experimental.AssetImporters.AssetImportContext ctx) (at Library/PackageCache/com.unity.barracuda@0.7.1-preview/Barracuda/Editor/ONNXModelImporter.cs:958)
UnityEditor.Experimental.AssetImporters.ScriptedImporter.GenerateAssetData (UnityEditor.Experimental.AssetImporters.AssetImportContext ctx) (at <ffcf2f9855fe46038192299154f37a45>:0)
UnityEditorInternal.InternalEditorUtility:ProjectWindowDrag(HierarchyProperty, Boolean)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)

Is it because unsqueeze  is not supported in barracuda?

**Comment by FlorentGuinier** [comment]
Hi Jaggernaut007,

Unsqueeze is supported on Barracuda, the important bit here is as you mentionned in the title of the bug `Only tensors of rank 4 or less are supported, but got rank 5`. Said overwise the maximum number of dimensions Barracuda support for a given tensor at the moment is 4. Meaning you will get the above error if Unsqueeze try to reshape a tensor to have a layout with 5 axis or more. Is it possible to avoid the situation in your network?

**Comment by jaggernaut007** [comment]
I cannot bypass that. Would need it to be rank 5. Would it be possible to let barracuda unsqueeze at rank 5?

**Comment by AlexRibard** [comment]
@jaggernaut007 would it be possible for you to share your .onnx model so we can have a closer look?
(It can be with randomly initialized weights).
Thanks!

**Comment by jaggernaut007** [comment]
@AlexRibard  Here are the models you would need. I have only tried to use the onnx model, not the tensorflow one.

Onnx: https://drive.google.com/open?id=1_o_6Gg6arMN4CqKEvaaNPQ-jI3fbcbIV
Tensorflow: https://drive.google.com/open?id=1RfhotsQy_XOSuTuV8KOAT1powHBxowEu

**Comment by AlexRibard** [comment]
Thanks!

**Comment by AlexRibard** [comment]
I am seeing a few part that might be problematic for importing the model:
* Slice and dynamic Transpose is not implemented atm, so this won't work unfortunately.
It doesn't seem trivial to re-write it however...
![image](https://user-images.githubusercontent.com/12990800/82827713-8fddcd80-9e7d-11ea-833e-b739869a6610.png)
* This seems to be were the 5d tensor is coming from. Looks like you are trying to apply softmax only on part of the tensor. Is there a way to rewrite it?
![image](https://user-images.githubusercontent.com/12990800/82827797-bb60b800-9e7d-11ea-8dee-18e44acb1cc6.png)

**Comment by jaggernaut007** [comment]
Okay let me see what I can do about this. Will try to find a work around.

**Comment by AlexRibard** [comment]
Let me know!
For the Softmax part, maybe you could get away with some clever element wise * masking.
So you first mask out the parts not needed for the softmax with -FloatMax. (that way they don't count for the softmax average and are set to 0)
Apply softmax. 
Apply inverse mask on input again and add the two results.

**Comment by AlexRibard** [comment]
For the split, can you use the slice operator?
https://github.com/onnx/onnx/blob/master/docs/Operators.md#Slice
So you slice the input multiple times (with different strides), apply transpose and group back into one tensor.
That might even work for the Softmax part

**Comment by jaggernaut007** [comment]
Let me try these out. I have actually picked up the model from anibali/margipose.
Would need to modify the model from that which was converted. 

I want to try another approach before diving into modyfying the model.
Was going to experiment with changing the opset version while exporting. Seems liek the slice operator is from the opset 11. Will try opset 8 and try importing. Hopefully all operations from onnx opset 8 will be compatible with the barracuda engine.

**Comment by AlexRibard** [comment]
Nice!
Yeah, as a rule of thumb, prefer opset9 as this has most coverage in Barracuda.

**Comment by jaggernaut007** [comment]
Well onnx couldnt find a way to use opset 8 for the conversion. Had to use opset9.
Will use this as the new standard model now.
barracuda still complains about the rank.

I am trying to replace split with slice. but seems like pytorch doesnt support the slice function!
ANy other suggestions of how to overcome this?

**Comment by AlexRibard** [comment]
https://discuss.pytorch.org/t/tensor-slice-in-pytorch/1449/4
I am guessing that gets exported as slice by onnx? Could be wrong however.

**Comment by jaggernaut007** [comment]
Nice! This could work. Let me get the update model onto barracuda
